File: rust/src/merge_sort.rs
fn merge(vec:&mut Vec<usize>, left:usize, mid: usize, right: usize) {
    let mut first_half = vec[left..mid+1].to_vec();
    let mut second_half = vec[mid+1..right+1].to_vec();
    first_half.reverse();
    second_half.reverse();
    for i in left..right+1 {
        // check if first half empty remove the begin from second half to the vector else complier panics 
        if first_half.is_empty() {
            vec[i] = second_half.pop().unwrap();
            continue; 
        }
        // likewise for the second half
        if second_half.is_empty() {
            vec[i] = first_half.pop().unwrap();
            continue; 
        }
        if second_half.last() < first_half.last() {
            vec[i] = second_half.pop().unwrap();
        } else {
            vec[i] = first_half.pop().unwrap();
        }
    }
}

pub fn merge_sort(vec: &mut Vec<usize>, left: usize, right: usize) {
    if left < right {
        let mid: usize = (left + right)/2;
        merge_sort(vec, left, mid);
        merge_sort(vec, mid+1, right);
        merge(vec, left, mid, right);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn merge_sort_test_1(){
        let mut vec = vec![4,9,2,8,1];
        merge_sort(&mut vec,0,4);
        assert_eq!(vec,[1,2,4,8,9]);
    }

    #[test]    
    fn merge_sort_test_2(){
        let mut vec = vec![4,9,2,8,1,54,44,43,84];
        merge_sort(&mut vec,0,8);
        assert_eq!(vec,[1,2,4,8,9,43,44,54,84]);
    }
}

File: c++/src/merge_sort.cpp
#include <iostream>
#include <merge_sort.h>

int *worker_array_addr;
static void merge_sort (int *p_array, int left, int right, int *p_worker_array);
static void merge_arrays (int *A, int l, int m, int r, int *B);

using namespace std;

void merge_sort_start (int *p_array, int size) {
	int worker_array[size];
	for (int i = 0; i < size; i++) {
		worker_array[i] = p_array[i];
	}
	/*cout << "first :: ";
	for ( int w = 0; w < size; w++) {
		cout << worker_array[w] << ", ";
	}
	cout << endl;*/
	worker_array_addr = (int *)worker_array;
	merge_sort(p_array, 0, (size-1), worker_array_addr);
}

static void merge_sort (int *p_array, int left, int right, int *p_worker_array) {
	
		// Find the middle point
		int mid = (left + right) / 2;
	if ((right - left) > 1) {
		//if (mid > left) {
			// Merge sort the first half
			merge_sort(p_array, left, (mid-1), p_worker_array);
		//}
		//if (mid < right) {
			// Merge sort the second half
			merge_sort(p_array, mid, right, p_worker_array);
		//}
		// Merge the two halves sorted in previous steps
		merge_arrays(p_array, left, mid, right, p_worker_array);
	}
	
	else if ((right - left) == 1){
		if (p_array[left] > p_array[right]) {
			int temp = p_array[left];
			p_array[left] = p_array[right];
			p_array[right] = temp;
		}
		for (int k = left; k <= right; k++)
			p_worker_array[k] = p_array[k];
	}
	
	return;
}

static void merge_arrays (int *A, int l, int m, int r, int *B) {
	int i = l, j = m;
	//cout << "Step :: ";
	for (int k = l; k <= r; k++) {
		if ((i < m) 
			&& ((j > r) || B[i] <= B[j])) 
		{
			A[k] = B[i];
			i++;
		} 
		else {
			A[k] = B[j];
			j++;
		}
		//cout << A[k] << ", ";
	}
	for (int k = l; k <= r; k++)
		B[k] = A[k];
	//cout << endl;
	/*
	cout << "L = " << l <<" M = " << m << " R = " << r << endl;
	cout << "Step :: ";
	for ( int w = l; w <= r; w++) {
		cout << A[w] << ", ";
	}
	cout << endl << "--------------------------" << endl;
	*/
}